### PR TITLE
feat: add controllers, tests and helpers for auth

### PR DESCRIPTION
**Pull request by karansapolia**
adds tests, middlewares and jwt helper files for auth.

fix: #66

## PR Checklist
Please check that your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/pesto-students/batch-10-tripen/wiki/Commit-message-guidelines)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] GIFs/Videos/Images added (compulsory for frontend changes)

## PR Type
What kind of change does this PR introduce? 

- [ ] Bugfix
- [x] Feature
- [ ] Tests changed / added / updated
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Infrastructure changes
- [ ] Dependency upgrade
- [ ] Other... Please describe:

## Issue Number: 
#66 

## What is the new behavior?
Adds controllers, basic tests and jwtHelper for /auth.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: server/src/controllers/v1/user.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    if (isMatch) {
+      const userData = destructureData(user);
+      const token = newToken(userData);
+      return res.status(200).json({
+        success: true,
+        message: 'Sign-in successful',
+        user: { ...userData, token },
+      });
+    }
+    return res.status(400).json({
```
Comment (rishabh1403) [inline server/src/controllers/v1/user.js]: 403 error for wrong creds

File: server/src/controllers/v1/user.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  } catch (error) {
+    return res.status(400).json({
+      sucess: false,
+      message: 'Email not found. Please sign up',
+    });
+  }
+};
+
+const signupController = async (req, res) => {
+  try {
```
Comment (rishabh1403) [inline server/src/controllers/v1/user.js]: add some line breaks, the code is not readable

File: server/src/controllers/v1/user.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      });
+    }
+    return res
+      .status(200)
+      .json({
+        success: true,
+        message: 'Please check your email for the reset password instructions',
+      });
+  } catch (err) {
+    return res.status(400).json({ success: true, message: 'Some internal error occurued', err });
```
Comment (rishabh1403) [inline server/src/controllers/v1/user.js]: success = true ??

File: server/src/middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      message: 'Missing some imput fields. Please fill the details.',
+    }, req, res);
+  }
+  return next();
+};
+
+export const signinValidation = async (req, res, next) => {
+  if (!req.body.email || !req.body.password) {
+    return errorMessage({
+      status: 404,
```
Comment (rishabh1403) [inline server/src/middlewares/auth.js]: 404 is not a correct code here. 
Your creds are missing, but route/ page is still intact

File: server/src/routes/api/v1/auth.js [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
-router.post('/signup', [/* signupValidation */], /* signupController */ (req, res) => {
-  res.status(201).json({ status: 'signup controller reached' });
-}); // validations & controllers TODO
-router.post('/forgot-password', [/* forgotpasswordValidation */], /* forgotpasswordController */ (req, res) => {
-  res.status(200).json({ status: 'forgot password controller reached' });
-}); // validations & controllers TODO
+router.get('/status', controllerStatus);
+router.post('/signin', signinValidation, signinController);
+router.post('/signup', signupValidation, signupController);
+router.post('/forgot-password', [/* forgotpasswordValidation */], forgotPasswordController);
```
Comment (rishabh1403) [inline server/src/routes/api/v1/auth.js]: please remove comments

File: server/src/configs/server.js [hunk 6]
```diff
@@ -1,42 +1,44 @@
...
-    const start = () => {
-        create();
-        const port = server.get('port');
-        application = server.listen(port, () => {
-            console.log(`Tripen server active on port ${port}`);
-        });
-    };
+  const start = () => {
+    create();
+    mongoose.connect(env.databaseURI, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true });
```
Comment (nimish-gupta) [inline server/src/configs/server.js]: mongoose.connect returns a promise, use async/await and also handle the cases in which connection is failed, you can add this in separate PR and then merge that PR and rebase this to new master and push again. Or you can do the changes in this PR.

File: server/src/configs/env.js [hunk 7]
```diff
@@ -5,6 +5,10 @@ const configs = {
   port: process.env.PORT || 3030,
   databaseURI: process.env.DATABASE_URI || 'mongodb://localhost:27017/tripen',
   clientUrl: process.env.CLIENT_URL || 'http://localhost:3000',
+  secrets: {
+    jwt: process.env.SECRET,
+    jwtExp: '5d',
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/configs/env.js]: I see no harm in using it in this way. Do you suggest keeping an independent secrets object?

File: server/src/controllers/v1/auth.test.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        name: 'Remula Pinto',
+        username: 'remu.pinto',
+        password: '1234',
+        location: {
+          type: 'Point',
+          coordinates: [1, 2],
+        },
+      };
+      const token = newToken(inputUser);
+      const user = await verifyToken(token);
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.test.js]: > in the above method you have not used await with verify token.
> Why the inconsistency?
> Same with controllers ?

That was an error. Added the await to upper method.

File: server/src/controllers/v1/auth.test.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+      expect(user.id).toBe(inputUser.id);
+    });
+  });
+
+  describe('signup: ', () => {
+    it('requires email and password', async () => {
+      expect.assertions(2);
+      const req = { body: {} };
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.test.js]: > what is this checking?
> Where are the assertions?

It's checking whether the controller throws an error if no required inputs are given in req.body.
The assertions are right below, in the status() and json() functions.

File: server/src/controllers/v1/auth.test.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        },
+        async json(result) {
+          let user = await verifyToken(result.token);
+          user = await User.findById(user.id)
+            .lean()
+            .exec();
+          expect(user.email).toBe('hello@hello.com');
+        },
+      };
+      await signupController(req, res);
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.test.js]: This is checking if correct token is being generated for each new registered user.

File: server/src/controllers/v1/auth.test.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+      const res = {
+        status(status) {
+          expect(status).toBe(400);
+          return this;
+        },
+        json(result) {
+          expect(typeof result.message).toBe('string');
+        },
+      };
+      await signinController(req, res);
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.test.js]: To throw an error if `req.body` does not include email or password inputs

File: server/src/controllers/v1/auth.js [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+const forgotPasswordController = async (req, res) => {
+  try {
+    const currentTime = new Date();
+    const expiryTime = new Date(currentTime.getTime() + 600 * 1000).valueOf();
+    const token = { value: newToken(), expires: expiryTime };
+    const user = await User.findOneAndUpadte({ email: req.body.email }).select('name email');
+    if (user) {
+      await sendMail({
+        type: 'forgot-password',
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.js]: Done.

File: server/src/routes/api/v1/auth.js (hunk 5 above)
Comment (karansapolia) [reply server/src/routes/api/v1/auth.js]: Done

File: server/src/controllers/v1/user.js (hunk 3 above)
Comment (karansapolia) [reply server/src/controllers/v1/user.js]: Good catch. Removed.

File: server/src/controllers/v1/user.js (hunk 2 above)
Comment (karansapolia) [reply server/src/controllers/v1/user.js]: Removed.

File: server/src/controllers/v1/user.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
+import user from '../../models/user';
+import destructureData from '../utils/index';
+import { newToken } from '../utils/jwtToken';
+
+const health = (req, res) => {
+  res.status(200).json({ status: 'ok' });
+};
+
+const signinController = async (req, res) => {
+  try {
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/user.js]: File removed from this PR.

File: server/src/controllers/v1/user.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    const user = await user.findOne({
+      email: req.body.email,
+    });
+    const isMatch = await user.checkPassword(req.body.password);
+    if (isMatch) {
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/user.js]: File removed from this PR.

File: server/src/controllers/v1/user.js (hunk 1 above)
Comment (karansapolia) [reply server/src/controllers/v1/user.js]: File removed from this PR.

File: server/src/middlewares/auth.js (hunk 4 above)
Comment (karansapolia) [reply server/src/middlewares/auth.js]: Corrected, thanks

File: server/src/controllers/v1/auth.js [hunk 15]
```diff
@@ -0,0 +1,105 @@
+import { sendMail } from 'nodemailer';
+import User from '../../models/user';
+import { newToken } from '../../utils/helpers/jwtHelper';
+import errorMessage from '../../utils/errorMessage';
+import statusMessage from '../../utils/statusMessage';
+
+const controllerStatus = (req, res) => res.status(200).json({ status: 'online' });
+
+const signinController = async (req, res) => {
+  try {
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.js]: [Implemented as requested](https://github.com/pesto-students/batch-10-tripen/blob/fea8ec751a76d39b572dce5f06409c3cf3aca0fe/server/src/controllers/v1/auth.js#L12)

File: server/src/controllers/v1/auth.js [hunk 16]
```diff
@@ -0,0 +1,105 @@
+import { sendMail } from 'nodemailer';
+import User from '../../models/user';
+import { newToken } from '../../utils/helpers/jwtHelper';
+import errorMessage from '../../utils/errorMessage';
+import statusMessage from '../../utils/statusMessage';
+
+const controllerStatus = (req, res) => res.status(200).json({ status: 'online' });
+
+const signinController = async (req, res) => {
```
[the remark this answers is not in the document]
Comment (karansapolia) [reply server/src/controllers/v1/auth.js]: Refactored to two try blocks.

File: server/src/controllers/v1/auth.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
+import User from '../../models/user';
+import { newToken } from '../../utils/helpers/jwtHelper';
+import errorMessage from '../../utils/errorMessage';
+import statusMessage from '../../utils/statusMessage';
+import destructureUser from '../../utils/destructureUser';
+
+const controllerStatus = (req, res) => res.status(200).json({ status: 'online' });
+
+const signinController = async (req, res) => {
+  let user;
```
Comment (rishabh1403) [inline server/src/controllers/v1/auth.js]: - make this a const
- move inside try, it has lexical scoping, so I don't see any issue with it, unless you have a reason to use it outside, I'm happy to discuss :) 
- if !user is shouldn't be thrown, that's expected behaviour, don't throw, return a response with message that account doesn't exist sign up
- if await throws , catch it in catch block, that should be intented behaviour of try/catch 

leme know if you have more doubts on this

apologoies for misspellings :D

File: server/src/controllers/v1/auth.js [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+      const userData = destructureUser(user);
+      const token = newToken({ ...userData });
+      return statusMessage({
+        success: true,
+        status: 200,
+        message: 'Sign-in successful',
+        user: { ...userData, token },
+      }, req, res);
+    }
+    throw Error;
```
Comment (rishabh1403) [inline server/src/controllers/v1/auth.js]: same as above

File: server/src/controllers/v1/auth.js [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    return errorMessage({
+      status: 400,
+      message: 'Sign-in failed. Check email and/or password',
+    }, req, res);
+  }
+};
+
+const signupController = async (req, res) => {
+  try {
+    const user = await User.create(req.body);
```
Comment (rishabh1403) [inline server/src/controllers/v1/auth.js]: line breaks please, it's very unreadbable

File: server/src/controllers/v1/auth.test.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (rishabh1403) [reply server/src/controllers/v1/auth.test.js]: but you are checking the token anywhere ,  or am i missing something here, 
you are calling signup controller, but not checkig it's result ? I'm little confused :smile:

File: server/src/utils/destructureUser.js [hunk 20]
```diff
@@ -0,0 +1,15 @@
+const destructureUser = ({
+  _id,
```
Comment (rishabh1403) [inline server/src/utils/destructureUser.js]: why not just destruct user using spread or pull properties when you need it ?

File: server/src/utils/validations/common/isEmail.js [hunk 21]
```diff
@@ -0,0 +1,13 @@
+import addErrorIfNotValid from '../addErrorIfNotValid';
+
+const isEmail = (email) => {
+  const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (rishabh1403) [inline server/src/utils/validations/common/isEmail.js]: did you write this regex? 
do you understand this regex ?

Comment (karansapolia) [reply server/src/utils/validations/common/isEmail.js]: Picked it up from [this website](https://emailregex.com/), tested it on regexr.com. I understand it. Covers all email cases I could find.

File: server/src/controllers/v1/auth.js (hunk 19 above)
Comment (karansapolia) [reply server/src/controllers/v1/auth.js]: ESLint automatically removes the first empty line after start of a block

**Pull request merged**